components/engine_traits/src/util.rs

```diff
@@ -21,3 +24,182 @@ pub fn check_key_in_range(
+///      migration is interrupted and reopened in the reverse direction.
+pub struct RaftDataStateMachine {
+    source: PathBuf,
+    source_staging: PathBuf,
+    target: PathBuf,
+    target_staging: PathBuf,
+}
+
+impl RaftDataStateMachine {
+    pub fn new(source: &str, target: &str) -> Self {
+        let source = PathBuf::from(source);
+        let source_staging = source.with_extension("STAGING");
+        let target = PathBuf::from(target);
+        let target_staging = target.with_extension("STAGING");
+        Self {
+            source,
+            source_staging,
+            target,
+            target_staging,
+        }
+    }
+
+    /// Checks if the current condition is a valid state.
+    pub fn validate(&self, should_exist: bool) -> std::result::Result<(), String> {
+        if Self::data_exists(&self.source) && Self::data_exists(&self.target) {
+            return Err(format!(
+                "Found multiple raft data sets: {}, {}",
+                self.source.display(),
+                self.target.display()
+            ));
+        }
+        if Self::data_exists(&self.source_staging) && Self::data_exists(&self.target_staging) {
+            return Err(format!(
+                "Found multiple raft data sets: {}, {}",
+                self.source_staging.display(),
+                self.target_staging.display()
+            ));
+        }
+        let exists = Self::data_exists(&self.source_staging)
+            || Self::data_exists(&self.target_staging)
+            || Self::data_exists(&self.source)
+            || Self::data_exists(&self.target);
+        if exists != should_exist {
+            if should_exist {
+                return Err("Cannot find raft data set.".to_owned());
+            } else {
+                return Err("Found raft data set when it should not exist.".to_owned());
+            }
+        }
+        Ok(())
+    }
+
+    /// Enters the `Migrating` state and returns the source directory if a
+    /// migration is needed. Otherwise prepares the target directory for
+    /// opening.
+    pub fn before_open_target(&mut self) -> Option<PathBuf> {
+        // Clean up trash directory if there is any.
+        for p in [
+            &self.source,
+            &self.source_staging,
+            &self.target,
+            &self.target_staging,
+        ] {
+            let trash = p.with_extension("REMOVE");
+            if trash.exists() {
+                fs::remove_dir_all(&trash).unwrap();
+            }
+        }
+        if Self::data_exists(&self.target_staging) {
+            // InvMigrating -> Completed
+            assert!(!Self::data_exists(&self.source_staging));
+            Self::remove_dir_safe(&self.source);
+            Self::rename_dir_safe(&self.target_staging, &self.target);
+            return None;
+        }
+        if Self::data_exists(&self.source_staging) {
+            // Double-check.
+            Self::remove_dir_safe(&self.source);
```

Comment:
How can this happen?

Reply:
It can't happen. But the protocol itself doesn't require `source` to be absent in this condition too. I'll make the requirement stronger and replace these loose ends with assertions.